File: Win8.Js.Advertising/js/FreeWheelPlugin.js
(function (PlayerFramework, undefined) {
    "use strict";

    // FreeWheelPlugin Errors
    var invalidConstruction = "Invalid construction: FreeWheelPlugin constructor must be called using the \"new\" operator.",
        companionAdFailed = "Companion ad failed.";

    var freeWheelTrackingEventArea = "freewheel";

    // FreeWheelPlugin Class
    var FreeWheelPlugin = WinJS.Class.derive(PlayerFramework.Plugins.AdSchedulerPlugin, function (options) {
        if (!(this instanceof PlayerFramework.Plugins.FreeWheelPlugin)) {
            throw invalidConstruction;
        }

        this._lastTrackingTime = null;

        PlayerFramework.Plugins.AdSchedulerPlugin.call(this, options);
    }, {
        // Public Properties
        source: {
            get: function () {
                return this._source;
            },
            set: function (value) {
                this._source = value;
            }
        },

        // Public Methods
        loadAds: function (source) {
            this._adSlots = [];

            var promise = Microsoft.VideoAdvertising.FreeWheelFactory.loadSource(new Windows.Foundation.Uri(source)).then(
                function (result) {
                    var adResponse = result;

                    var videoTracking = PlayerFramework.Utilities.first(adResponse.siteSection.videoPlayer.videoAsset.eventCallbacks, function (eventCallback) {
                        return eventCallback.name === Microsoft.VideoAdvertising.FWEventCallback.videoView;
                    });

                    if (videoTracking) {
                        this._lastTrackingTime = null;

                        // position tracking
                        this.mediaPlayer.positionTrackingPlugin.trackingEvents.push({ positionPercentage: 0, data: videoTracking, area: freeWheelTrackingEventArea });
                        this.mediaPlayer.positionTrackingPlugin.trackingEvents.push({ positionPercentage: 1, data: videoTracking, area: freeWheelTrackingEventArea });

                        // play time tracking
                        for (var i = 15; i < 60; i += 15) {
                            this.mediaPlayer.playTimeTrackingPlugin.trackingEvents.push({ playTime: i, data: videoTracking, area: freeWheelTrackingEventArea });
                        }

                        for (var i = 60; i < 60 * 3; i += 30) {
                            this.mediaPlayer.playTimeTrackingPlugin.trackingEvents.push({ playTime: i, data: videoTracking, area: freeWheelTrackingEventArea });
                        }

                        for (var i = 60 * 3; i < 60 * 10; i += 60) {
                            this.mediaPlayer.playTimeTrackingPlugin.trackingEvents.push({ playTime: i, data: videoTracking, area: freeWheelTrackingEventArea });
                        }

                        for (var i = 60 * 10; i < 60 * 30; i += 120) {
                            this.mediaPlayer.playTimeTrackingPlugin.trackingEvents.push({ playTime: i, data: videoTracking, area: freeWheelTrackingEventArea });
                        }

                        for (var i = 60 * 30; i < 60 * 60; i += 300) {
                            this.mediaPlayer.playTimeTrackingPlugin.trackingEvents.push({ playTime: i, data: videoTracking, area: freeWheelTrackingEventArea });
                        }

                        for (var i = 60 * 60; i < 60 * 180; i += 600) {
                            this.mediaPlayer.playTimeTrackingPlugin.trackingEvents.push({ playTime: i, data: videoTracking, area: freeWheelTrackingEventArea });
                        }
                    }

                    var videoAsset = adResponse.siteSection.videoPlayer.videoAsset;

                    var promises = null;
                    if (videoAsset) {
                        promises = [];
                        for (var i = 0; i < videoAsset.adSlots.length; i++) {
                            var adSlot = videoAsset.adSlots[i];
                            (function(adSlot){
                                promises.push(Microsoft.VideoAdvertising.FreeWheelFactory.getAdDocumentPayload(adSlot, adResponse).then(function (payload) {
                                        var ad = null;
                                        switch (adSlot.timePositionClass) {
                                            case "preroll":
                                                ad = new PlayerFramework.Advertising.PrerollAdvertisement();
                                                break;

                                            case "postroll":
                                                ad = new PlayerFramework.Advertising.PostrollAdvertisement();
                                                break;

                                            default:
                                                ad = new PlayerFramework.Advertising.MidrollAdvertisement();
                                                ad.time = adSlot.timePosition / 1000;
                                                break;
                                        }

                                        ad.source = new Microsoft.PlayerFramework.Js.Advertising.AdSource();
                                        ad.source.type = Microsoft.VideoAdvertising.DocumentAdPayloadHandler.adType;
                                        ad.source.payload = payload;

                                        this.advertisements.push(ad);
                                        this._adSlots.push({ "ad": ad, "adSlot": adSlot });
                                }.bind(this), function () { /* ignore */ }));
                            }).bind(this)(adSlot);
                        }
                        this._adResponse = adResponse;
                    }

                    this._loadCompanions();
                    if (promises) {
                        return WinJS.Promise.join(promises);
                    }
                }.bind(this)
            );

            promise.done(
                function () {
                    PlayerFramework.Utilities.remove(this._activePromises, promise);
                }.bind(this),
                function (e) {
                    PlayerFramework.Utilities.remove(this._activePromises, promise);
                }.bind(this)
            );

            this._activePromises.push(promise);
            return promise;
        },

        // Private Methods
        _setOptions: function (options) {
            PlayerFramework.Utilities.setOptions(this, options, {
                isEnabled: true,
                advertisements: [],
                preloadTime: 5,
                evaluateOnForwardOnly: true,
                interruptScrub: true,
                source: null
            });
        },

        _getAdSlot: function (ad) {
            var item = PlayerFramework.Utilities.first(this._adSlots, function (item) {
                return item.ad === ad;
            });

            if (item) {
                return item.adSlot;
            }

            return null;
        },

        _playAd: function (ad) {
            var adSlot = this._getAdSlot(ad);

            if (adSlot) {
                var impression = PlayerFramework.Utilities.first(adSlot.eventCallbacks, function (eventCallback) {
                    return eventCallback.type === Microsoft.VideoAdvertising.FWCallbackType.impression && eventCallback.name === Microsoft.VideoAdvertising.FWEventCallback.slotImpression;
                });

                if (impression) {
                    var urls = impression.getUrls();
                    for (var i = 0; i < urls.length; i++) {
                        var url = urls[i];
                        Microsoft.VideoAdvertising.AdTracking.current.fireTracking(url);
                    }
                }
            }
            
            return PlayerFramework.Plugins.AdSchedulerPlugin.prototype._playAd.call(this, ad);
        },

        _loadCompanions: function () {
            if (this._adResponse) {
                var enumerable = Microsoft.VideoAdvertising.FreeWheelFactory.getNonTemporalCompanions(this._adResponse);
                this._companions = PlayerFramework.Utilities.getArray(enumerable);
                this.mediaPlayer.adHandlerPlugin.loadCompanions(this._companions, Microsoft.VideoAdvertising.CompanionAdsRequired.all)
            }
        },

        _unloadCompanions: function () {
            if (this._companions) {
                this.mediaPlayer.adHandlerPlugin.unloadCompanions(this._companions);
                this._companions = null;
            }
        },

        _getTrackingUrl: function (url, start, end) {
            var now = Date.now();
            var init = start ? 1 : 0;
            var last = end ? 1 : 0;
            var ct = this._lastTrackingTime ? Math.round((now - this._lastTrackingTime) / 1000) : 0;
            
            // save for next time
            this._lastTrackingTime = now;

            if (url.indexOf("?") !== -1) {
                return PlayerFramework.Utilities.formatString("{0}&init={1}&last={2}&ct={3}", url, init, last, ct);
            } else {
                return PlayerFramework.Utilities.formatString("{0}?init={1}&last={2}&ct={3}", url, init, last, ct);
            }
        },

        _onLoad: function () {
            this._adSlots = [];
            this._adResponse = null;
            this._companions = null;

            PlayerFramework.Plugins.AdSchedulerPlugin.prototype._onLoad.call(this);
        },

        _onUnload: function () {
            this._adSlots = null;
            this._adResponse = null;
            this._companions = null;

            PlayerFramework.Plugins.AdSchedulerPlugin.prototype._onUnload.call(this);
        },

        _onActivate: function () {
            this._bindEvent("loading", this.mediaPlayer, this._onMediaPlayerLoading);
            this._bindEvent("eventtracked", this.mediaPlayer.playTimeTrackingPlugin, this._onMediaPlayerPlayTimeEventTracked);
            this._bindEvent("eventtracked", this.mediaPlayer.positionTrackingPlugin, this._onMediaPlayerPositionEventTracked);

            this._loadCompanions();

            return PlayerFramework.Plugins.AdSchedulerPlugin.prototype._onActivate.call(this);
        },

        _onDeactivate: function () {
            this._unloadCompanions();

            this._unbindEvent("loading", this.mediaPlayer, this._onMediaPlayerLoading);
            this._unbindEvent("eventtracked", this.mediaPlayer.playTimeTrackingPlugin, this._onMediaPlayerPlayTimeEventTracked);
            this._unbindEvent("eventtracked", this.mediaPlayer.positionTrackingPlugin, this._onMediaPlayerPositionEventTracked);

            PlayerFramework.Plugins.AdSchedulerPlugin.prototype._onDeactivate.call(this);
        },

        _onUpdate: function () {
            this._unloadCompanions();

            this._adSlots = [];
            this._adResponse = null;
            this._companions = null;

            PlayerFramework.Plugins.AdSchedulerPlugin.prototype._onUpdate.call(this);
        },

        _onMediaPlayerLoading: function (e) {
            if (this.source) {
                var promise = this.loadAds(this.source);
                e.detail.setPromise(promise);
            }
        },

        _onMediaPlayerPlayTimeEventTracked: function (e) {
            var trackingEvent = e.detail.trackingEvent;
            if (trackingEvent.area === freeWheelTrackingEventArea) {
                if (trackingEvent.data instanceof Microsoft.VideoAdvertising.FWEventCallback) {
                    var urls = PlayerFramework.Utilities.getArray(trackingEvent.data.getUrls());
                    for (var i = 0; i < urls.length; i++) {
                        var url = urls[i];
                        var trackingUrl = this._getTrackingUrl(url);
                        Microsoft.VideoAdvertising.AdTracking.current.fireTracking(trackingUrl);
                    }
                }
            }
        },

        _onMediaPlayerPositionEventTracked: function (e) {
            var trackingEvent = e.detail.trackingEvent;
            if (trackingEvent.area === freeWheelTrackingEventArea) {
                if (trackingEvent.data instanceof Microsoft.VideoAdvertising.FWEventCallback) {
                    var urls = PlayerFramework.Utilities.getArray(trackingEvent.data.getUrls());
                    for (var i = 0; i < urls.length; i++) {
                        var url = urls[i];
                        var start = !isNaN(trackingEvent.positionPercentage) && trackingEvent.positionPercentage === 0;
                        var end = !isNaN(trackingEvent.positionPercentage) && trackingEvent.positionPercentage === 1;
                        var trackingUrl = this._getTrackingUrl(url, start, end);
                        Microsoft.VideoAdvertising.AdTracking.current.fireTracking(trackingUrl);
                    }
                }
            }
        }
    });

    // FreeWheelPlugin Mixins
    WinJS.Class.mix(PlayerFramework.MediaPlayer, {
        freeWheelPlugin: {
            value: null,
            writable: true,
            configurable: true
        }
    });

    // FreeWheelPlugin Exports
    WinJS.Namespace.define("PlayerFramework.Plugins", {
        FreeWheelPlugin: FreeWheelPlugin
    });

})(PlayerFramework);

File: Win8.Xaml.Advertising/FreeWheelPlugin.cs
using System;
using System.Linq;
using System.IO;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.VideoAdvertising;
using System.Collections.Generic;
#if SILVERLIGHT
using System.Windows;
#else
using Windows.UI.Xaml;
#endif

namespace Microsoft.PlayerFramework.Advertising
{
    /// <summary>
    /// A plugin that is capable of downloading a FreeWheel SmartXML source file, parsing it and using it to schedule when ads should play.
    /// </summary>
    public class FreeWheelPlugin : AdSchedulerPlugin
    {
        readonly Dictionary<Advertisement, FWTemporalAdSlot> adSlots = new Dictionary<Advertisement, FWTemporalAdSlot>();
        private CancellationTokenSource cts;
        private FWAdResponse adResponse;
        private bool trackingEnded;
        private PlayTimeTrackingEvent lastTrackingEvent;
        const string TrackingEventArea = "FreeWheel";

        /// <summary>
        /// Identifies the Source dependency property.
        /// </summary>
        public static readonly DependencyProperty SourceProperty = DependencyProperty.Register("Source", typeof(Uri), typeof(FreeWheelPlugin), null);

        /// <summary>
        /// Gets or sets the source Uri of the FreeWheel SmartXML file
        /// </summary>
        public Uri Source
        {
            get { return GetValue(SourceProperty) as Uri; }
            set { SetValue(SourceProperty, value); }
        }

        /// <inheritdoc /> 
        protected override void OnUpdate()
        {
            Source = FreeWheel.GetSource((DependencyObject)CurrentMediaSource);
            base.OnUpdate();
        }

        /// <inheritdoc /> 
        protected override bool OnActivate()
        {
            cts = new CancellationTokenSource();
            WirePlayer();
            if (adResponse != null) ShowCompanions();
            return base.OnActivate();
        }

        /// <inheritdoc /> 
        protected override void OnDeactivate()
        {
            MediaPlayer.UnloadAllCompanions();
            cts.Cancel();
            cts = null;
            UnwirePlayer();
            base.OnDeactivate();
        }

        private void WirePlayer()
        {
#if !WINDOWS_PHONE7
            MediaPlayer.MediaLoading += mediaPlayer_MediaLoading;
#endif
            MediaPlayer.MediaClosed += MediaPlayer_MediaClosed;
        }

        private void UnwirePlayer()
        {
#if !WINDOWS_PHONE7
            MediaPlayer.MediaLoading -= mediaPlayer_MediaLoading;
#endif
            MediaPlayer.MediaClosed -= MediaPlayer_MediaClosed;
        }

        private void trackingPlugin_EventTracked(object sender, EventTrackedEventArgs e)
        {
            if (e.TrackingEvent.Area == TrackingEventArea)
            {
                var eventCallback = e.TrackingEvent.Data as FWEventCallback;
                if (eventCallback != null)
                {
                    foreach (var url in eventCallback.GetUrls())
                    {
                        TrackVideoViewMarker(url, e);
                    }
                }
            }
        }

        private void TrackVideoViewMarker(string url, EventTrackedEventArgs e)
        {
            if (!trackingEnded)
            {
                bool isStart;
                bool isEnd;
                TimeSpan currentPlayTime;
                if (e.TrackingEvent is PositionTrackingEvent)
                {
                    var playTimeTrackingPlugin = MediaPlayer.Plugins.OfType<PlayTimeTrackingPlugin>().FirstOrDefault();
                    if (playTimeTrackingPlugin == null) throw new Exception("PlayTimeTrackingPlugin not found; required for FreeWheelPlugin");
                    var positionTrackingEvent = (PositionTrackingEvent)e.TrackingEvent;
                    if (!positionTrackingEvent.PositionPercentage.HasValue || positionTrackingEvent.PositionPercentage.Value != 1) throw new Exception("Invalid tracking event was registered for FreeWheelPlugin");
                    isStart = false;
                    isEnd = true;
                    currentPlayTime = playTimeTrackingPlugin.PlayTime;
                    trackingEnded = true; // set this flag to prevent further tracking
                }
                else if (e.TrackingEvent is PlayTimeTrackingEvent)
                {
                    var playTimeTrackingEvent = (PlayTimeTrackingEvent)e.TrackingEvent;
                    isStart = playTimeTrackingEvent.PlayTime == TimeSpan.Zero;
                    isEnd = false;
                    currentPlayTime = playTimeTrackingEvent.PlayTime;
                }
                else
                {
                    throw new ArgumentException();
                }

                TimeSpan delta;
                if (lastTrackingEvent != null)
                {
                    delta = currentPlayTime - lastTrackingEvent.PlayTime;
                }
                else
                {
                    delta = currentPlayTime;
                }
                lastTrackingEvent = e.TrackingEvent as PlayTimeTrackingEvent;

                var newUrl = url + string.Format("{3}init={0}&ct={1}&last={2}", isStart ? 1 : 0, (int)Math.Round(delta.TotalSeconds), isEnd ? 1 : 0, url.Contains("?") ? "&" : "?");
                AdTracking.Current.FireTracking(newUrl);
            }
        }

#if WINDOWS_PHONE7
        protected override async void MediaPlayer_MediaLoading(object sender, MediaPlayerDeferrableEventArgs e)
#else
        async void mediaPlayer_MediaLoading(object sender, MediaPlayerDeferrableEventArgs e)
#endif
        {
            if (IsEnabled)
            {
                if (Source != null)
                {
                    var deferral = e.DeferrableOperation.GetDeferral();
                    try
                    {
                        await LoadAds(Source, deferral.CancellationToken);
#if WINDOWS_PHONE7
                        await base.PlayStartupAds(deferral.CancellationToken);
#endif
                    }
                    catch { /* ignore */ }
                    finally
                    {
                        deferral.Complete();
                    }
                }
            }
        }

        /// <summary>
        /// Loads ads from a source URI. Note, this is called automatically if you set the source before the MediaLoading event fires and normally does not need to be called.
        /// </summary>
        /// <param name="source">The SmartXML source URI</param>
        /// <param name="c">A cancellation token that allows you to cancel a pending operation</param>
        /// <returns>A task to await on.</returns>
        public async Task LoadAds(Uri source, CancellationToken c)
        {
            adSlots.Clear();
            var cancellationToken = CancellationTokenSource.CreateLinkedTokenSource(c, cts.Token).Token;

#if SILVERLIGHT
            adResponse = await FreeWheelFactory.LoadSource(source, cancellationToken);
#else
            adResponse = await FreeWheelFactory.LoadSource(source).AsTask(cancellationToken);
#endif

            var videoTracking = adResponse.SiteSection.VideoPlayer.VideoAsset.EventCallbacks.FirstOrDefault(ec => ec.Name == FWEventCallback.VideoView);
            if (videoTracking != null)
            {
                // use the tracking plugins to help with tracking markers. Create it if it doesn't exist.
                var positionTrackingPlugin = MediaPlayer.Plugins.OfType<PositionTrackingPlugin>().FirstOrDefault();
                if (positionTrackingPlugin == null)
                {
                    positionTrackingPlugin = new PositionTrackingPlugin();
                    MediaPlayer.Plugins.Add(positionTrackingPlugin);
                }
                positionTrackingPlugin.EventTracked += trackingPlugin_EventTracked;
                lastTrackingEvent = null; // reset
                trackingEnded = false;
                positionTrackingPlugin.TrackingEvents.Add(new PositionTrackingEvent() { PositionPercentage = 1, Data = videoTracking, Area = TrackingEventArea });

                var playTimeTrackingPlugin = MediaPlayer.Plugins.OfType<PlayTimeTrackingPlugin>().FirstOrDefault();
                if (playTimeTrackingPlugin == null)
                {
                    playTimeTrackingPlugin = new PlayTimeTrackingPlugin();
                    MediaPlayer.Plugins.Add(playTimeTrackingPlugin);
                }
                playTimeTrackingPlugin.EventTracked += trackingPlugin_EventTracked;
                for (int i = 0; i < 60; i = i + 15)
                    playTimeTrackingPlugin.TrackingEvents.Add(new PlayTimeTrackingEvent() { PlayTime = TimeSpan.FromSeconds(i), Data = videoTracking, Area = TrackingEventArea });
                for (int i = 60; i < 60 * 3; i = i + 30)
                    playTimeTrackingPlugin.TrackingEvents.Add(new PlayTimeTrackingEvent() { PlayTime = TimeSpan.FromSeconds(i), Data = videoTracking, Area = TrackingEventArea });
                for (int i = 60 * 3; i < 60 * 10; i = i + 60)
                    playTimeTrackingPlugin.TrackingEvents.Add(new PlayTimeTrackingEvent() { PlayTime = TimeSpan.FromSeconds(i), Data = videoTracking, Area = TrackingEventArea });
                for (int i = 60 * 10; i < 60 * 30; i = i + 120)
                    playTimeTrackingPlugin.TrackingEvents.Add(new PlayTimeTrackingEvent() { PlayTime = TimeSpan.FromSeconds(i), Data = videoTracking, Area = TrackingEventArea });
                for (int i = 60 * 30; i < 60 * 60; i = i + 300)
                    playTimeTrackingPlugin.TrackingEvents.Add(new PlayTimeTrackingEvent() { PlayTime = TimeSpan.FromSeconds(i), Data = videoTracking, Area = TrackingEventArea });
                for (int i = 60 * 60; i < 60 * 180; i = i + 600)
                    playTimeTrackingPlugin.TrackingEvents.Add(new PlayTimeTrackingEvent() { PlayTime = TimeSpan.FromSeconds(i), Data = videoTracking, Area = TrackingEventArea });
            }

            var videoAsset = adResponse.SiteSection.VideoPlayer.VideoAsset;
            if (videoAsset != null)
            {
                foreach (var adSlot in videoAsset.AdSlots)
                {
                    Advertisement ad = null;
                    switch (adSlot.TimePositionClass)
                    {
                        case "preroll":
                            ad = new PrerollAdvertisement();
                            break;
                        case "postroll":
                            ad = new PostrollAdvertisement();
                            break;
                        default:
                            var midroll = new MidrollAdvertisement();
                            midroll.Time = adSlot.TimePosition;
                            ad = midroll;
                            break;
                    }

#if SILVERLIGHT
                    var payload = await FreeWheelFactory.GetAdDocumentPayload(adSlot, adResponse, cancellationToken);
#else
                    var payload = await FreeWheelFactory.GetAdDocumentPayload(adSlot, adResponse).AsTask(cancellationToken);
#endif
                    ad.Source = new AdSource(payload, DocumentAdPayloadHandler.AdType);

                    Advertisements.Add(ad);
                    adSlots.Add(ad, adSlot);
                }
            }

            ShowCompanions();
        }

        private void MediaPlayer_MediaClosed(object sender, RoutedEventArgs e)
        {
            var playTimeTrackingPlugin = MediaPlayer.Plugins.OfType<PlayTimeTrackingPlugin>().FirstOrDefault();
            if (playTimeTrackingPlugin != null)
            {
                playTimeTrackingPlugin.EventTracked -= trackingPlugin_EventTracked;
            }

            var positionTrackingPlugin = MediaPlayer.Plugins.OfType<PlayTimeTrackingPlugin>().FirstOrDefault();
            if (positionTrackingPlugin != null)
            {
                positionTrackingPlugin.EventTracked -= trackingPlugin_EventTracked;
            }

            adSlots.Clear();
        }

        private void ShowCompanions()
        {
            foreach (var companionCreative in FreeWheelFactory.GetNonTemporalCompanions(adResponse))
            {
                MediaPlayer.ShowCompanion(companionCreative);
            }
        }

        /// <inheritdoc /> 
        protected override async Task PlayAdAsync(Advertisement ad, CancellationToken cancellationToken)
        {
            if (adSlots.ContainsKey(ad)) // app could have manually added ads besides those from FreeWheel
            {
                var adSlot = adSlots[ad];
                try
                {
                    var slotImpression = adSlot.EventCallbacks.FirstOrDefault(ec => ec.Type == FWCallbackType.Impression && ec.Name == FWEventCallback.SlotImpression);
                    if (slotImpression != null)
                    {
                        foreach (var url in slotImpression.GetUrls())
                        {
                            AdTracking.Current.FireTracking(url);
                        }
                    }
                    if (ad.Source != null)
                    {
                        await base.PlayAdAsync(ad, cancellationToken);
                    }
                }
                catch { /* swallow */ }
            }
            else
            {
                await base.PlayAdAsync(ad, cancellationToken);
            }
        }

    }
}
